File: EX1/GIT_GROUP_20/calculator.java
import java.math.BigDecimal;
import java.math.BigInteger;
import java.util.Scanner;

public class calculator {
    /*
     * Lab notes:
     * 1. The program supports positive and negative numbers.
     * 2. The program supports whole numbers and fractional numbers.
     * 3. Spaces at the beginning or end are ignored, but spaces inside the
     *    number are not allowed.
     * 4. Hex input may use 0-9, A-F, or a-f.
     * 5. Hex output is always printed with uppercase letters.
     * 6. A number may contain only one decimal point.
     * 7. The input must contain at least one digit.
     * 8. When converting from decimal to hex, the fractional part is limited
     *    to 8 hex digits after the point. This keeps the output simple and
     *    avoids infinite repeating fractions.
     * 9. When converting from hex to decimal, the result is exact for any
     *    finite hex fraction.
     */

    private static final int MAX_HEX_FRACTION_DIGITS = 8;
    private static final BigInteger BIG_INTEGER_SIXTEEN = BigInteger.valueOf(16);
    private static final BigDecimal BIG_DECIMAL_SIXTEEN = BigDecimal.valueOf(16);

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        // Short introduction so the user knows what the program does.
        System.out.println("-------------------------------------------------");
        System.out.println("Welcome to Group 20 Base 10 / Base 16 Calculator!");
        System.out.println("-------------------------------------------------");

        while (true) {
            printMenu();
            String choice = scanner.nextLine().trim();

            if (choice.equals("0")) {
                System.out.println("Program ended.");
                break;
            }

            if (choice.equals("1")) {
                handleDecimalToHex(scanner);
            } else if (choice.equals("2")) {
                handleHexToDecimal(scanner);
            } else {
                System.out.println("Error: Please enter 1, 2, or 0.");
            }

            System.out.println();
        }

        scanner.close();
    }

    private static void printMenu() {
        // The menu is kept simple because this is meant for a beginner-friendly lab.
        System.out.println();
        System.out.println("Choose conversion:");
        System.out.println("1. Base 10 to Base 16");
        System.out.println("2. Base 16 to Base 10");
        System.out.println("0. Exit");
        System.out.print("Your choice: ");
    }

    private static void handleDecimalToHex(Scanner scanner) {
        // Read one line from the user, validate it, then convert it.
        System.out.print("Enter a decimal number: ");
        String input = scanner.nextLine();

        try {
            ParsedNumber parsedNumber = parseDecimalInput(input);
            String hexValue = decimalToHex(parsedNumber);
            System.out.println("Result: " + hexValue);
        } catch (IllegalArgumentException error) {
            System.out.println(error.getMessage());
        }
    }

    private static void handleHexToDecimal(Scanner scanner) {
        // Read one line from the user, validate it, then convert it.
        System.out.print("Enter a hexadecimal number: ");
        String input = scanner.nextLine();

        try {
            ParsedNumber parsedNumber = parseHexInput(input);
            String decimalValue = hexToDecimal(parsedNumber);
            System.out.println("Result: " + decimalValue);
        } catch (IllegalArgumentException error) {
            System.out.println(error.getMessage());
        }
    }

    private static ParsedNumber parseDecimalInput(String rawInput) {
        // First split the text into sign, integer part, and fractional part.
        ParsedTextParts parts = splitAndValidate(rawInput, false);

        // Check that every character in the integer part is a decimal digit.
        for (int index = 0; index < parts.integerPart.length(); index++) {
            char currentChar = parts.integerPart.charAt(index);
            if (!Character.isDigit(currentChar)) {
                throw new IllegalArgumentException(
                    "Error: Invalid decimal character '" + currentChar + "'."
                );
            }
        }

        // Check that every character in the fractional part is also a decimal digit.
        for (int index = 0; index < parts.fractionalPart.length(); index++) {
            char currentChar = parts.fractionalPart.charAt(index);
            if (!Character.isDigit(currentChar)) {
                throw new IllegalArgumentException(
                    "Error: Invalid decimal character '" + currentChar + "'."
                );
            }
        }

        BigInteger integerValue = parseUnsignedDecimalInteger(parts.integerPart);
        BigDecimal fractionalValue = parseDecimalFraction(parts.fractionalPart);

        return new ParsedNumber(parts.negative, integerValue, fractionalValue);
    }

    private static ParsedNumber parseHexInput(String rawInput) {
        // First split the text into sign, integer part, and fractional part.
        ParsedTextParts parts = splitAndValidate(rawInput, true);

        // Check that every character in the integer part is a valid hex digit.
        for (int index = 0; index < parts.integerPart.length(); index++) {
            char currentChar = parts.integerPart.charAt(index);
            if (hexDigitValue(currentChar) == -1) {
                throw new IllegalArgumentException(
                    "Error: Invalid hexadecimal character '" + currentChar + "'."
                );
            }
        }

        // Check that every character in the fractional part is also a valid hex digit.
        for (int index = 0; index < parts.fractionalPart.length(); index++) {
            char currentChar = parts.fractionalPart.charAt(index);
            if (hexDigitValue(currentChar) == -1) {
                throw new IllegalArgumentException(
                    "Error: Invalid hexadecimal character '" + currentChar + "'."
                );
            }
        }

        BigInteger integerValue = parseUnsignedHexInteger(parts.integerPart);
        BigDecimal fractionalValue = parseHexFraction(parts.fractionalPart);

        return new ParsedNumber(parts.negative, integerValue, fractionalValue);
    }

    private static ParsedTextParts splitAndValidate(String rawInput, boolean hexadecimal) {
        // Remove spaces around the number. Example: "  26.75  " becomes "26.75".
        String input = rawInput.trim();

        if (input.isEmpty()) {
            throw new IllegalArgumentException(
                hexadecimal
                    ? "Error: Hexadecimal input cannot be empty."
                    : "Error: Decimal input cannot be empty."
            );
        }

        boolean negative = false;
        int startIndex = 0;
        char firstChar = input.charAt(0);

        if (firstChar == '+' || firstChar == '-') {
            negative = firstChar == '-';
            startIndex = 1;
        }

        // Reject inputs like "+" or "-" with no digits after the sign.
        if (startIndex == input.length()) {
            throw new IllegalArgumentException(
                hexadecimal
                    ? "Error: Hexadecimal input must contain digits after the sign."
                    : "Error: Decimal input must contain digits after the sign."
            );
        }

        String unsignedPart = input.substring(startIndex);
        int pointCount = 0;
        boolean hasDigit = false;

        for (int index = 0; index < unsignedPart.length(); index++) {
            char currentChar = unsignedPart.charAt(index);

            // Only one decimal point is allowed.
            if (currentChar == '.') {
                pointCount++;
                if (pointCount > 1) {
                    throw new IllegalArgumentException("Error: Input cannot contain more than one decimal point.");
                }
                continue;
            }

            // A sign in the middle is always an error.
            if (currentChar == '+' || currentChar == '-') {
                throw new IllegalArgumentException("Error: Sign characters are allowed only at the beginning.");
            }

            // Internal spaces are not accepted because they make the format unclear.
            if (Character.isWhitespace(currentChar)) {
                throw new IllegalArgumentException("Error: Spaces inside the number are not allowed.");
            }

            hasDigit = true;
        }

        if (!hasDigit) {
            throw new IllegalArgumentException(
                hexadecimal
                    ? "Error: Hexadecimal input must contain at least one digit."
                    : "Error: Decimal input must contain at least one digit."
            );
        }

        int pointIndex = unsignedPart.indexOf('.');
        String integerPart;
        String fractionalPart;

        if (pointIndex == -1) {
            integerPart = unsignedPart;
            fractionalPart = "";
        } else {
            integerPart = unsignedPart.substring(0, pointIndex);
            fractionalPart = unsignedPart.substring(pointIndex + 1);
        }

        // Inputs like ".5" are allowed, so we treat the missing integer part as 0.
        if (integerPart.isEmpty()) {
            integerPart = "0";
        }

        return new ParsedTextParts(negative, integerPart, fractionalPart);
    }

    private static BigInteger parseUnsignedDecimalInteger(String digits) {
        // Build the decimal integer manually from left to right.
        // Example: "255" -> ((0 * 10 + 2) * 10 + 5) * 10 + 5
        BigInteger result = BigInteger.ZERO;

        for (int index = 0; index < digits.length(); index++) {
            int digit = digits.charAt(index) - '0';
            result = result.multiply(BigInteger.TEN).add(BigInteger.valueOf(digit));
        }

        return result;
    }

    private static BigDecimal parseDecimalFraction(String digits) {
        if (digits.isEmpty()) {
            return BigDecimal.ZERO;
        }

        // Example: "75" means 75 / 100 = 0.75
        BigInteger numerator = BigInteger.ZERO;
        BigInteger denominator = BigInteger.ONE;

        for (int index = 0; index < digits.length(); index++) {
            int digit = digits.charAt(index) - '0';
            numerator = numerator.multiply(BigInteger.TEN).add(BigInteger.valueOf(digit));
            denominator = denominator.multiply(BigInteger.TEN);
        }

        return new BigDecimal(numerator).divide(new BigDecimal(denominator));
    }

    private static BigInteger parseUnsignedHexInteger(String digits) {
        // Build the integer part using powers of 16.
        // Example: "1C" = (1 * 16) + 12
        BigInteger result = BigInteger.ZERO;
        BigInteger power = BigInteger.ONE;

        for (int index = digits.length() - 1; index >= 0; index--) {
            int digit = hexDigitValue(digits.charAt(index));
            result = result.add(power.multiply(BigInteger.valueOf(digit)));
            power = power.multiply(BIG_INTEGER_SIXTEEN);
        }

        return result;
    }

    private static BigDecimal parseHexFraction(String digits) {
        // Build the fractional part using negative powers of 16.
        // Example: ".8" = 8 / 16 = 0.5
        // Example: ".F" = 15 / 16 = 0.9375
        BigDecimal result = BigDecimal.ZERO;
        BigDecimal divisor = BIG_DECIMAL_SIXTEEN;

        for (int index = 0; index < digits.length(); index++) {
            int digit = hexDigitValue(digits.charAt(index));

            if (digit != 0) {
                BigDecimal digitValue = BigDecimal.valueOf(digit);
                result = result.add(digitValue.divide(divisor));
            }

            divisor = divisor.multiply(BIG_DECIMAL_SIXTEEN);
        }

        return result;
    }

    private static String decimalToHex(ParsedNumber value) {
        // Handle zero early so we do not print "-0" or "0.".
        if (value.isZero()) {
            return "0";
        }

        String integerHex = convertIntegerPartToHex(value.integerPart);
        String fractionalHex = convertFractionPartToHex(value.fractionalPart);
        StringBuilder builder = new StringBuilder();

        if (value.negative) {
            builder.append('-');
        }

        builder.append(integerHex);

        if (!fractionalHex.isEmpty()) {
            builder.append('.').append(fractionalHex);
        }

        return builder.toString();
    }

    private static String convertIntegerPartToHex(BigInteger integerPart) {
        if (integerPart.equals(BigInteger.ZERO)) {
            return "0";
        }

        // Repeated division by 16 gives the hex digits from right to left.
        StringBuilder builder = new StringBuilder();
        BigInteger currentValue = integerPart;

        while (!currentValue.equals(BigInteger.ZERO)) {
            BigInteger[] divisionResult = currentValue.divideAndRemainder(BIG_INTEGER_SIXTEEN);
            int remainder = divisionResult[1].intValue();
            builder.append(hexCharacter(remainder));
            currentValue = divisionResult[0];
        }

        return builder.reverse().toString();
    }

    private static String convertFractionPartToHex(BigDecimal fractionalPart) {
        if (fractionalPart.compareTo(BigDecimal.ZERO) == 0) {
            return "";
        }

        StringBuilder builder = new StringBuilder();
        BigDecimal currentValue = fractionalPart;

        /*
         * Repeated multiplication by 16 gives one digit at a time after the point.
         * Example:
         * 0.75 * 16 = 12.0, so the first hex digit is C.
         * We stop after MAX_HEX_FRACTION_DIGITS digits so repeating values do
         * not continue forever.
         */
        for (int index = 0; index < MAX_HEX_FRACTION_DIGITS; index++) {
            currentValue = currentValue.multiply(BIG_DECIMAL_SIXTEEN);
            int digit = currentValue.intValue();
            builder.append(hexCharacter(digit));
            currentValue = currentValue.subtract(BigDecimal.valueOf(digit));

            if (currentValue.compareTo(BigDecimal.ZERO) == 0) {
                break;
            }
        }

        return removeTrailingZeroHexDigits(builder.toString());
    }

    private static String removeTrailingZeroHexDigits(String digits) {
        // If the fractional result ends with zeros, remove them.
        // Example: "C000" becomes "C"
        int endIndex = digits.length();

        while (endIndex > 0 && digits.charAt(endIndex - 1) == '0') {
            endIndex--;
        }

        return digits.substring(0, endIndex);
    }

    private static String hexToDecimal(ParsedNumber value) {
        // Decimal value = integer part + fractional part, then apply the sign.
        BigDecimal decimalValue = new BigDecimal(value.integerPart).add(value.fractionalPart);

        if (value.negative && decimalValue.compareTo(BigDecimal.ZERO) != 0) {
            decimalValue = decimalValue.negate();
        }

        return formatDecimal(decimalValue);
    }

    private static String formatDecimal(BigDecimal value) {
        // Clean the output so 28.5000 becomes 28.5 and 10.0000 becomes 10.
        BigDecimal normalizedValue = value.stripTrailingZeros();

        if (normalizedValue.compareTo(BigDecimal.ZERO) == 0) {
            return "0";
        }

        return normalizedValue.toPlainString();
    }

    private static int hexDigitValue(char character) {
        // Convert one hex character into its numeric value.
        // If the character is not valid, return -1.
        if (character >= '0' && character <= '9') {
            return character - '0';
        }

        if (character >= 'A' && character <= 'F') {
            return character - 'A' + 10;
        }

        if (character >= 'a' && character <= 'f') {
            return character - 'a' + 10;
        }

        return -1;
    }

    private static char hexCharacter(int value) {
        // Convert a number from 0 to 15 into one hex character.
        if (value >= 0 && value <= 9) {
            return (char) ('0' + value);
        }

        return (char) ('A' + (value - 10));
    }

    private static class ParsedTextParts {
        // This class stores the raw text pieces after the input is split.
        private final boolean negative;
        private final String integerPart;
        private final String fractionalPart;

        private ParsedTextParts(boolean negative, String integerPart, String fractionalPart) {
            this.negative = negative;
            this.integerPart = integerPart;
            this.fractionalPart = fractionalPart;
        }
    }

    private static class ParsedNumber {
        // This class stores the numeric value after the text has been parsed.
        private final boolean negative;
        private final BigInteger integerPart;
        private final BigDecimal fractionalPart;

        private ParsedNumber(boolean negative, BigInteger integerPart, BigDecimal fractionalPart) {
            this.negative = negative;
            this.integerPart = integerPart;
            this.fractionalPart = fractionalPart;
        }

        private boolean isZero() {
            return integerPart.equals(BigInteger.ZERO) && fractionalPart.compareTo(BigDecimal.ZERO) == 0;
        }
    }
}

File: EX1/GIT_GROUP_29/calculator.py
# Exercise A - Git
# Note: No built-in conversion functions are used
# (no hex(), int() with base, format(), oct(), bin())

import sys

HEX_DIGITS = "0123456789ABCDEF"


def parse_decimal_string(s):
    """
    Manually parses a decimal number string (integer or fraction)
    into integer_part and fractional_part as integers, plus the
    number of fractional digits.
    """

    dot_count = 0
    for ch in s:
        if ch == '.':
            dot_count += 1

    # If there is more than one decimal point, the number is invalid
    if dot_count > 1:
        raise ValueError("Invalid number: more than one decimal point.")

    if dot_count == 0:
        # Integer only
        for ch in s:
            if ch < '0' or ch > '9':
                raise ValueError(f"Invalid decimal digit: '{ch}'.")
        int_part = 0
        for ch in s:
            int_part = int_part * 10 + (ord(ch) - ord('0'))
        return int_part, 0, 0

    else:
        # Has decimal point
        parts = s.split('.')
        int_str = parts[0]
        frac_str = parts[1]

        # Validate integer part
        for ch in int_str:
            if ch < '0' or ch > '9':
                raise ValueError(f"Invalid decimal digit: '{ch}'.")
        # Validate fractional part
        for ch in frac_str:
            if ch < '0' or ch > '9':
                raise ValueError(f"Invalid decimal digit: '{ch}'.")

        int_part = 0
        if int_str != "":
            for ch in int_str:
                int_part = int_part * 10 + (ord(ch) - ord('0'))

        frac_part = 0
        frac_digits = len(frac_str)
        if frac_str != "":
            for ch in frac_str:
                frac_part = frac_part * 10 + (ord(ch) - ord('0'))

        return int_part, frac_part, frac_digits


def decimal_to_hex(decimal_str):
    """
    Converts a decimal (base 10) number string to hexadecimal (base 16).
    """

    # Remove whitespace to avoid invalid characters during manual validation
    decimal_str = decimal_str.strip()

    if decimal_str == "":
        raise ValueError("Input is empty.")

    # Handle negative sign
    negative = False
    if decimal_str[0] == '-':
        negative = True
        decimal_str = decimal_str[1:]
    elif decimal_str[0] == '+':
        decimal_str = decimal_str[1:]

    if decimal_str == "":
        raise ValueError("Input contains only a sign with no digits.")

    # Parse the decimal number manually
    int_part, frac_part, frac_digits = parse_decimal_string(decimal_str)

    # Convert integer part to hex by repeated division by 16
    if int_part == 0:
        hex_int = "0"
    else:
        hex_int = ""
        temp = int_part
        while temp > 0:
            remainder = temp % 16
            hex_int = HEX_DIGITS[remainder] + hex_int
            temp = temp // 16

    # Convert fractional part to hex by repeated multiplication by 16
    hex_frac = ""
    if frac_part > 0 and frac_digits > 0:
        # Build the fractional value as numerator / denominator
        numerator = frac_part
        denominator = 1
        for _ in range(frac_digits):
            denominator = denominator * 10

        max_frac_hex = 10  # limit fractional hex digits
        for _ in range(max_frac_hex):
            numerator = numerator * 16
            digit = numerator // denominator
            hex_frac = hex_frac + HEX_DIGITS[digit]
            numerator = numerator - digit * denominator
            if numerator == 0:
                break

    # Build final result
    if hex_frac != "":
        result = hex_int + "." + hex_frac
    else:
        result = hex_int

    if negative and result != "0":
        result = "-" + result

    return result


def hex_digit_value(ch):
    """Returns the numeric value of a hex digit (0-15)."""
    for i in range(16):
        if HEX_DIGITS[i] == ch:
            return i
    return -1


def hex_to_decimal(hex_str):
    """
    Converts a hexadecimal (base 16) string to decimal (base 10).
    """
    # Remove whitespace to avoid invalid characters during manual validation
    hex_str = hex_str.strip()

    if hex_str == "":
        raise ValueError("Input is empty.")

    # Handle negative sign
    negative = False
    if hex_str[0] == '-':
        negative = True
        hex_str = hex_str[1:]
    elif hex_str[0] == '+':
        hex_str = hex_str[1:]

    # Remove optional 0x / 0X prefix
    if len(hex_str) >= 2 and hex_str[0] == '0' and (hex_str[1] == 'x' or hex_str[1] == 'X'):
        hex_str = hex_str[2:]

    if hex_str == "":
        raise ValueError("Input contains no hexadecimal digits.")

    # Convert to uppercase for uniform processing
    hex_str = hex_str.upper()

    # Split into integer and fractional parts
    dot_count = 0
    for ch in hex_str:
        if ch == '.':
            dot_count += 1
    if dot_count > 1:
        raise ValueError("Invalid number: more than one decimal point.")

    if dot_count == 0:
        int_str = hex_str
        frac_str = ""
    else:
        parts = hex_str.split('.')
        int_str = parts[0]
        frac_str = parts[1]

    # Validate integer part
    for ch in int_str:
        if ch not in HEX_DIGITS:
            raise ValueError(f"Invalid hexadecimal digit: '{ch}'.")

    # Validate fractional part
    for ch in frac_str:
        if ch not in HEX_DIGITS:
            raise ValueError(f"Invalid hexadecimal digit: '{ch}'.")

    # Convert integer part: multiply by 16 for each digit
    int_result = 0
    for ch in int_str:
        int_result = int_result * 16 + hex_digit_value(ch)

    # Convert fractional part using exact arithmetic (numerator / denominator)
    if frac_str == "":
        # Integer only
        result_str = str_from_int(int_result)
    else:
        frac_numerator = 0
        frac_denominator = 1
        for ch in frac_str:
            frac_numerator = frac_numerator * 16 + hex_digit_value(ch)
            frac_denominator = frac_denominator * 16

        # Compute fractional decimal digits
        # frac_value = frac_numerator / frac_denominator
        # Multiply to get decimal digits
        frac_decimal = ""
        temp_num = frac_numerator
        temp_den = frac_denominator
        max_digits = 10
        for _ in range(max_digits):
            temp_num = temp_num * 10
            digit = temp_num // temp_den
            frac_decimal = frac_decimal + str_from_int(digit)
            temp_num = temp_num - digit * temp_den
            if temp_num == 0:
                break

        # Remove trailing zeros from fractional part
        while len(frac_decimal) > 1 and frac_decimal[-1] == '0':
            frac_decimal = frac_decimal[:-1]

        if frac_decimal == "" or frac_decimal == "0":
            result_str = str_from_int(int_result)
        else:
            result_str = str_from_int(int_result) + "." + frac_decimal

    if negative and result_str != "0":
        result_str = "-" + result_str

    return result_str


def str_from_int(n):
    """Convert a non-negative integer to string manually."""
    if n == 0:
        return "0"
    result = ""
    temp = n
    while temp > 0:
        digit = temp % 10
        result = HEX_DIGITS[digit] + result
        temp = temp // 10
    return result


def main():
    """
    Main interactive menu for the base converter calculator.
    """
    print("=" * 55)
    print("     Base Converter Calculator")
    print("     Hexadecimal <-> Decimal")
    print("=" * 55)

    while True:
        print("\n1. Decimal -> Hexadecimal")
        print("2. Hexadecimal -> Decimal")
        print("3. Exit")
        sys.stdout.flush()
        choice = input("Choose (1-3): ").strip()

        # Convert decimal input to hexadecimal
        if choice == "1":
            sys.stdout.flush()
            dec_input = input("Enter decimal number: ").strip()
            try:
                result = decimal_to_hex(dec_input)
                print("Result: " + result)
            except ValueError as e:
                print("Error:", e)
            sys.stdout.flush()

        # Convert hexadecimal input to decimal
        elif choice == "2":
            sys.stdout.flush()
            hex_input = input("Enter hexadecimal number: ").strip()
            try:
                result = hex_to_decimal(hex_input)
                print("Result: " + result)
            except ValueError as e:
                print("Error:", e)
            sys.stdout.flush()

        elif choice == "3":
            print("Goodbye!")
            break

        else:
            print("Invalid option. Please choose 1-3.")


if __name__ == "__main__":
    main()